File: ui/src/lib/sitemap.ts
import type { Request } from '@/api/client'

function is2xx(req: Request): boolean {
  const status = req.response?.status_code
  return status != null && status >= 200 && status < 300
}

export interface SitemapBranchNode {
  id: string
  kind: 'host' | 'segment'
  label: string
  fullPath: string
  requestCount: number
  responseCount: number
  children: SitemapNode[]
}

export interface SitemapRequestNode {
  id: string
  kind: 'request'
  request: Request
  occurrenceCount: number
  responseCount: number
}

export type SitemapNode = SitemapBranchNode | SitemapRequestNode

interface MutableBranchNode {
  id: string
  kind: 'host' | 'segment'
  label: string
  fullPath: string
  requestCount: number
  responseCount: number
  branches: Map<string, MutableBranchNode>
  leaves: Map<string, SitemapRequestNode>
}

function createBranch(kind: 'host' | 'segment', id: string, label: string, fullPath: string): MutableBranchNode {
  return {
    id,
    kind,
    label,
    fullPath,
    requestCount: 0,
    responseCount: 0,
    branches: new Map(),
    leaves: new Map(),
  }
}

function sortNodes(nodes: SitemapNode[]): SitemapNode[] {
  return [...nodes].sort((a, b) => {
    if (a.kind === 'request' && b.kind === 'request') return b.request.id - a.request.id
    if (a.kind === 'request') return 1
    if (b.kind === 'request') return -1
    return a.label.localeCompare(b.label, undefined, { numeric: true, sensitivity: 'base' })
  })
}

function finalize(node: MutableBranchNode): SitemapBranchNode {
  const branchChildren = Array.from(node.branches.values()).map(finalize)
  const children = sortNodes([...branchChildren, ...Array.from(node.leaves.values())])

  return {
    id: node.id,
    kind: node.kind,
    label: node.label,
    fullPath: node.fullPath,
    requestCount: node.requestCount,
    responseCount: node.responseCount,
    children,
  }
}

export function buildSitemapTree(requests: Request[]): SitemapBranchNode[] {
  const hosts = new Map<string, MutableBranchNode>()

  for (const request of requests) {
    const hostKey = request.host || 'unknown-host'
    let hostNode = hosts.get(hostKey)
    if (!hostNode) {
      hostNode = createBranch('host', `host:${hostKey}`, hostKey, hostKey)
      hosts.set(hostKey, hostNode)
    }

    const parts = request.path
      .split('/')
      .filter(Boolean)
      .map((part) => part || '/')

    let cursor = hostNode
    let currentPath = ''

    for (const part of parts) {
      currentPath += `/${part}`
      let branch = cursor.branches.get(part)
      if (!branch) {
        branch = createBranch('segment', `${cursor.id}:${currentPath}`, part, currentPath)
        cursor.branches.set(part, branch)
      }
      cursor = branch
    }

    const routeKey = `${request.scheme}://${request.host}${request.path || '/'}`
    const existingLeaf = cursor.leaves.get(routeKey)
    if (existingLeaf) {
      existingLeaf.occurrenceCount += 1
      if (request.response) existingLeaf.responseCount += 1
      const existingIs2xx = is2xx(existingLeaf.request)
      const newIs2xx = is2xx(request)
      const shouldReplace =
        (!existingIs2xx && newIs2xx) ||
        (existingIs2xx === newIs2xx && request.id > existingLeaf.request.id)
      if (shouldReplace) {
        existingLeaf.request = request
        existingLeaf.id = `request:${request.id}`
      }
      continue
    }

    cursor.leaves.set(routeKey, {
      id: `request:${request.id}`,
      kind: 'request',
      request,
      occurrenceCount: 1,
      responseCount: request.response ? 1 : 0,
    })
  }

  function bubbleCounts(node: MutableBranchNode): { requestCount: number; responseCount: number } {
    let requestCount = node.leaves.size
    let responseCount = Array.from(node.leaves.values()).filter((leaf) => leaf.responseCount > 0).length

    for (const child of node.branches.values()) {
      const counts = bubbleCounts(child)
      requestCount += counts.requestCount
      responseCount += counts.responseCount
    }

    node.requestCount = requestCount
    node.responseCount = responseCount
    return { requestCount, responseCount }
  }

  for (const host of hosts.values()) {
    bubbleCounts(host)
  }

  return sortNodes(Array.from(hosts.values()).map(finalize)) as SitemapBranchNode[]
}

export function countUniqueRoutes(requests: Request[]): number {
  return new Set(requests.map((request) => `${request.scheme}://${request.host}${request.path || '/'}`)).size
}

export function getDefaultExpanded(tree: SitemapBranchNode[]): string[] {
  return tree.map((node) => node.id)
}

export function collectRequestIdsUnder(nodes: SitemapNode[]): number[] {
  const ids: number[] = []

  function visit(node: SitemapNode) {
    if (node.kind === 'request') {
      ids.push(node.request.id)
    } else {
      for (const child of node.children) visit(child)
    }
  }

  for (const node of nodes) visit(node)
  return ids
}

export function collectBranchIds(nodes: SitemapBranchNode[]): Set<string> {
  const ids = new Set<string>()

  function visit(node: SitemapBranchNode) {
    ids.add(node.id)
    for (const child of node.children) {
      if (child.kind !== 'request') visit(child)
    }
  }

  for (const node of nodes) visit(node)
  return ids
}

File: internal/mcp/sitemap.go
// SPDX-License-Identifier: Apache-2.0
package mcp

import (
	"fmt"
	"sort"
	"strings"

	proj "github.com/hamedsj5/pandorabox/internal/project"
	"github.com/hamedsj5/pandorabox/internal/proxy"
	"github.com/hamedsj5/pandorabox/internal/storage"
)

type sitemapBranchNode struct {
	ID            string        `json:"id"`
	Kind          string        `json:"kind"`
	Label         string        `json:"label"`
	FullPath      string        `json:"full_path"`
	RequestCount  int           `json:"request_count"`
	ResponseCount int           `json:"response_count"`
	Children      []interface{} `json:"children"`
}

type sitemapRequestNode struct {
	ID              string           `json:"id"`
	Kind            string           `json:"kind"`
	Request         *storage.Request `json:"request"`
	OccurrenceCount int              `json:"occurrence_count"`
	ResponseCount   int              `json:"response_count"`
}

type mutableBranchNode struct {
	ID            string
	Kind          string
	Label         string
	FullPath      string
	RequestCount  int
	ResponseCount int
	Branches      map[string]*mutableBranchNode
	Leaves        map[string]*sitemapRequestNode
}

func buildSitemapTree(requests []*storage.Request) []sitemapBranchNode {
	hosts := map[string]*mutableBranchNode{}

	for _, request := range requests {
		hostKey := request.Host
		if hostKey == "" {
			hostKey = "unknown-host"
		}

		hostNode, ok := hosts[hostKey]
		if !ok {
			hostNode = createBranch("host", "host:"+hostKey, hostKey, hostKey)
			hosts[hostKey] = hostNode
		}

		parts := splitPath(request.Path)
		cursor := hostNode
		currentPath := ""
		for _, part := range parts {
			currentPath += "/" + part
			branch, ok := cursor.Branches[part]
			if !ok {
				branch = createBranch("segment", fmt.Sprintf("%s:%s", cursor.ID, currentPath), part, currentPath)
				cursor.Branches[part] = branch
			}
			cursor = branch
		}

		routeKey := fmt.Sprintf("%s://%s%s", request.Scheme, request.Host, defaultPath(request.Path))
		existing, ok := cursor.Leaves[routeKey]
		if ok {
			existing.OccurrenceCount++
			if request.Response != nil {
				existing.ResponseCount++
			}
			if shouldReplaceLeaf(existing.Request, request) {
				existing.Request = request
				existing.ID = fmt.Sprintf("request:%d", request.ID)
			}
			continue
		}

		cursor.Leaves[routeKey] = &sitemapRequestNode{
			ID:              fmt.Sprintf("request:%d", request.ID),
			Kind:            "request",
			Request:         request,
			OccurrenceCount: 1,
			ResponseCount:   boolToInt(request.Response != nil),
		}
	}

	for _, host := range hosts {
		bubbleCounts(host)
	}

	nodes := make([]sitemapBranchNode, 0, len(hosts))
	for _, host := range hosts {
		nodes = append(nodes, finalizeBranch(host))
	}
	sort.Slice(nodes, func(i, j int) bool {
		return strings.Compare(strings.ToLower(nodes[i].Label), strings.ToLower(nodes[j].Label)) < 0
	})
	return nodes
}

func createBranch(kind, id, label, fullPath string) *mutableBranchNode {
	return &mutableBranchNode{
		ID:       id,
		Kind:     kind,
		Label:    label,
		FullPath: fullPath,
		Branches: map[string]*mutableBranchNode{},
		Leaves:   map[string]*sitemapRequestNode{},
	}
}

func finalizeBranch(node *mutableBranchNode) sitemapBranchNode {
	children := make([]interface{}, 0, len(node.Branches)+len(node.Leaves))
	for _, branch := range node.Branches {
		finalized := finalizeBranch(branch)
		children = append(children, finalized)
	}
	for _, leaf := range node.Leaves {
		children = append(children, *leaf)
	}
	sort.Slice(children, func(i, j int) bool {
		leftReq, leftIsReq := children[i].(sitemapRequestNode)
		rightReq, rightIsReq := children[j].(sitemapRequestNode)
		if leftIsReq && rightIsReq {
			return leftReq.Request.ID > rightReq.Request.ID
		}
		if leftIsReq {
			return false
		}
		if rightIsReq {
			return true
		}
		leftBranch := children[i].(sitemapBranchNode)
		rightBranch := children[j].(sitemapBranchNode)
		return strings.Compare(strings.ToLower(leftBranch.Label), strings.ToLower(rightBranch.Label)) < 0
	})

	return sitemapBranchNode{
		ID:            node.ID,
		Kind:          node.Kind,
		Label:         node.Label,
		FullPath:      node.FullPath,
		RequestCount:  node.RequestCount,
		ResponseCount: node.ResponseCount,
		Children:      children,
	}
}

func bubbleCounts(node *mutableBranchNode) (int, int) {
	requestCount := len(node.Leaves)
	responseCount := 0
	for _, leaf := range node.Leaves {
		if leaf.ResponseCount > 0 {
			responseCount++
		}
	}
	for _, child := range node.Branches {
		reqCount, respCount := bubbleCounts(child)
		requestCount += reqCount
		responseCount += respCount
	}
	node.RequestCount = requestCount
	node.ResponseCount = responseCount
	return requestCount, responseCount
}

func splitPath(path string) []string {
	parts := strings.Split(path, "/")
	filtered := make([]string, 0, len(parts))
	for _, part := range parts {
		if part != "" {
			filtered = append(filtered, part)
		}
	}
	return filtered
}

func defaultPath(path string) string {
	if path == "" {
		return "/"
	}
	return path
}

func shouldReplaceLeaf(existing, next *storage.Request) bool {
	existing2xx := has2xx(existing)
	next2xx := has2xx(next)
	if !existing2xx && next2xx {
		return true
	}
	return existing2xx == next2xx && next.ID > existing.ID
}

func has2xx(req *storage.Request) bool {
	return req != nil && req.Response != nil && req.Response.StatusCode >= 200 && req.Response.StatusCode < 300
}

func countUniqueRoutes(requests []*storage.Request) int {
	seen := map[string]struct{}{}
	for _, request := range requests {
		seen[fmt.Sprintf("%s://%s%s", request.Scheme, request.Host, defaultPath(request.Path))] = struct{}{}
	}
	return len(seen)
}

func countResponses(requests []*storage.Request) int {
	count := 0
	for _, request := range requests {
		if request.Response != nil {
			count++
		}
	}
	return count
}

func filterInScopeRequests(requests []*storage.Request, cfg proj.ScopeConfig) []*storage.Request {
	scope := &proxy.ScopeChecker{}
	scope.SetConfig(cfg)
	filtered := make([]*storage.Request, 0, len(requests))
	for _, request := range requests {
		if scope.InScope(request.Host, request.Path) {
			filtered = append(filtered, request)
		}
	}
	return filtered
}

func boolToInt(v bool) int {
	if v {
		return 1
	}
	return 0
}
